File: ui/js/pagefilter.js
// Search inside the page you are looking at.
//
// Not the header's search box, which spans the whole library and takes you to
// a different screen. This one narrows what is already in front of you —
// thirty-five consoles, twenty-seven collections, two and a half thousand
// arcade games — and leaves you where you are.
//
// It lived inside the collections list, drawn as part of that pane, which is
// why it only existed there and why it sat in the middle of a page of cards
// looking like something that had fallen out of a dialog. It is furniture of
// the tab row now: one box, always in the same place, filtering whatever the
// row is pointing at.

import { el, invoke } from "./state.js";

/// Hidden by the filter. A class rather than the `hidden` attribute so nothing
/// else that hides things — a view emptying itself, a card with no artwork —
/// gets confused with a search that did not match.
const OUT = "filtered-out";

/// What a row or card is called, for matching.
///
/// Every list here puts its name in one of these, and the fallback is the
/// node's own text — which catches the ones that do not, at the cost of also
/// matching a game's platform or its size. That is the right way round: a
/// filter that misses things is worse than one that is generous.
function nameOf(node) {
  const named = node.querySelector(".name, .nm, .hist-name");
  return (named?.textContent ?? node.textContent ?? "").toLowerCase();
}

function targets() {
  const region = el.consoles && !el.consoles.hidden ? [el.consoles, el.list] : [el.list];
  return region.flatMap((r) => [...r.querySelectorAll(".card, .gcard, .row, .prow, .hist-game")]);
}

let current = "";

/// Apply `text` to what is on screen. Empty text puts everything back.
///
/// What counts as a match, and when a group heading has been left with nothing
/// under it, are decided in `src/pagefilter.rs`. This finds the nodes, hands
/// over their names, and puts the class on whatever did not survive.
export async function applyPageFilter(text) {
  current = String(text ?? "").trim().toLowerCase();
  const nodes = targets();
  const heads = [...el.list.querySelectorAll(".ghead")];
  const at = new Map(nodes.map((n, i) => [n, i]));
  // Which entries sit under each heading, so a heading with nothing left under
  // it can go too: a search that leaves five headings and no games reads as a
  // broken page.
  const groups = heads.map((head) =>
    [...(head.parentElement?.querySelectorAll(".card, .row, .gcard") ?? [])]
      .map((n) => at.get(n))
      .filter((i) => i !== undefined)
  );

  const { visible, headings, shown } = await invoke("page_filter", {
    names: nodes.map(nameOf),
    query: current,
    groups,
  });
  nodes.forEach((node, i) => node.classList.toggle(OUT, !visible[i]));
  heads.forEach((head, i) => head.classList.toggle(OUT, !!headings[i]));
  return shown;
}

/// Re-apply after a list is redrawn. A redraw builds fresh nodes, which have
/// never seen the filter — without this, changing the order or coming back to
/// a tab quietly undoes the search still sitting in the box.
export function refreshPageFilter() {
  if (current) return applyPageFilter(current);
}

export function pageFilterText() {
  return current;
}

/// Empty the box and put everything back. Called when the view changes: a
/// filter typed for one screen means nothing on the next, and leaving the text
/// there means arriving at a list that is missing things for no visible
/// reason.
export function clearPageFilter() {
  current = "";
  if (el.pageFilter) el.pageFilter.value = "";
  for (const node of targets()) node.classList.remove(OUT);
  for (const head of el.list.querySelectorAll(".ghead")) head.classList.remove(OUT);
}

/// Say what this page holds, so the box can say what it will search.
export function setPageFilterLabel(what) {
  if (el.pageFilter) el.pageFilter.placeholder = what ? `Filter ${what}…` : "Filter this page…";
}

/// The slot beside the box, for the button that orders the list.
///
/// Collections have one; a console list does not. Whatever is put here is
/// replaced wholesale on the next view, so a view that says nothing gets an
/// empty slot rather than the last view's button.
export function setPageFilterExtra(node) {
  const slot = document.getElementById("page-filter-extra");
  if (!slot) return null;
  slot.replaceChildren();
  if (node) slot.appendChild(node);
  return slot;
}

export function installPageFilter() {
  if (!el.pageFilter) return;
  let timer;
  el.pageFilter.addEventListener("input", () => {
    clearTimeout(timer);
    // A beat behind the typing: on 2,506 games, filtering on every keystroke
    // is 2,506 class toggles per letter.
    timer = setTimeout(() => applyPageFilter(el.pageFilter.value), 120);
  });
  el.pageFilter.addEventListener("keydown", (ev) => {
    if (ev.key === "Escape") {
      clearPageFilter();
      el.pageFilter.blur();
    }
    // The list below owns the arrows; this box owns its own text.
    ev.stopPropagation();
  });
}

File: src/pagefilter.rs
// Search inside the page you are looking at.
//
// Not the library-wide search, which spans everything and takes you to a
// different screen. This one narrows what is already in front of you —
// thirty-five consoles, twenty-seven collections, two and a half thousand
// arcade games — and leaves you where you are.
//
// Two rules, and both of them are about not producing a page that looks
// broken.

/// Does this entry survive the text typed into the box?
///
/// Generous on purpose. Case and surrounding space are ignored, and the match
/// is anywhere in the name rather than at the start, because a filter that
/// misses things is worse than one that is too generous: a miss looks like the
/// game is not in the library.
///
/// Empty text keeps everything, which is how clearing the box puts the page
/// back.
pub fn matches(name: &str, query: &str) -> bool {
    let query = normalise(query);
    query.is_empty() || name.to_lowercase().contains(&query)
}

/// The typed text, as it is compared.
///
/// Exposed because callers hold the query across many rows and normalising it
/// once per row on 2,506 games is 2,506 allocations per keystroke.
pub fn normalise(query: &str) -> String {
    query.trim().to_lowercase()
}

/// Which of `names` survive.
pub fn visible(names: &[String], query: &str) -> Vec<bool> {
    let query = normalise(query);
    if query.is_empty() {
        return vec![true; names.len()];
    }
    names.iter().map(|n| n.to_lowercase().contains(&query)).collect()
}

/// Whether each group heading should go too.
///
/// A search that leaves five console headings and no games under any of them
/// reads as a broken page rather than as a search with no results. `groups`
/// holds the indices into `visible` that sit under each heading.
///
/// A heading with no members at all stays: an empty group is the page's own
/// business, and hiding it would make an unfiltered page differ from itself.
pub fn empty_groups(groups: &[Vec<usize>], visible: &[bool], query: &str) -> Vec<bool> {
    if normalise(query).is_empty() {
        return vec![false; groups.len()];
    }
    groups
        .iter()
        .map(|g| !g.is_empty() && !g.iter().any(|&i| visible.get(i).copied().unwrap_or(false)))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn names() -> Vec<String> {
        ["Super Mario Bros", "Metroid", "SUPER METROID", "Zelda"]
            .iter()
            .map(|s| (*s).to_owned())
            .collect()
    }

    fn kept(query: &str) -> Vec<String> {
        let names = names();
        visible(&names, query)
            .into_iter()
            .zip(&names)
            .filter(|(v, _)| *v)
            .map(|(_, n)| n.clone())
            .collect()
    }

    /// Ported from `ui/test/pagefilter.test.js`.
    #[test]
    fn it_keeps_what_matches_and_hides_the_rest() {
        assert_eq!(kept("metroid"), ["Metroid", "SUPER METROID"]);
    }

    #[test]
    fn case_does_not_matter_and_neither_does_stray_space() {
        assert_eq!(kept("  MeTrOiD  "), ["Metroid", "SUPER METROID"]);
    }

    #[test]
    fn emptying_it_puts_everything_back() {
        assert_eq!(kept("").len(), 4);
        assert_eq!(kept("   ").len(), 4, "a box holding only spaces hid the page");
    }

    /// A filter that misses things is worse than one that is too generous: a
    /// miss reads as the game not being in the library at all.
    #[test]
    fn it_matches_anywhere_in_the_name_not_only_the_start() {
        assert_eq!(kept("mario"), ["Super Mario Bros"]);
    }

    /// A search that leaves five headings and no games reads as a broken page.
    #[test]
    fn a_heading_with_nothing_left_under_it_goes_too() {
        let names = names();
        let groups = vec![vec![0], vec![1, 2], vec![3]];
        let seen = visible(&names, "metroid");
        assert_eq!(empty_groups(&groups, &seen, "metroid"), [true, false, true]);
    }

    /// An empty group is the page's own business. Hiding it would make an
    /// unfiltered page differ from itself.
    #[test]
    fn an_unfiltered_page_hides_no_headings() {
        let names = names();
        let groups = vec![vec![0], vec![], vec![1, 2, 3]];
        let seen = visible(&names, "");
        assert_eq!(empty_groups(&groups, &seen, ""), [false, false, false]);
    }
}
